Add Storage Service DIP integration

jraddaoui opened the pull request:
Allow to receive notifications when DIPs are stored in the SS to fetch
their METS files and create a Folder to represent it in the application.

- Add SS_HOSTS setting allowing multiple RFC-1738 SS URLs.
- Modify DIP model to include SS information.
- Create DIP from SS notification in DIPStoredWebhook:
  - Check Origin header against SS_HOSTS.
  - Get SS data, create DIP and launch async. tasks chain.
  - Split existing `extract_and_parse_mets` task and add `download_mets`
    and `save_import_error` tasks to ease both DIP import workflows.
  - Add import error to DIP model instead of tracking the TaskResult.
- Try to create DIP - Collection relation on METS parsing process.
- Proxy DIP from SS in download DIP view, prioritizing the local file
  if it exists.

Due to this integration, a Folder (DIP) may now not end related with a
Collection, which required a few changes:

- Allow null values in DIP's model collection field.
- Check that a relation with a Collection exists where needed.
- Do not require collection field on DIP creation form.
- Allow to modify/clear collection field on DIP edit form.
- Update Collection data in the Elasticsearch documents of related
  DigitalFiles when a DIP is modified.
- Add orphan folders page to show Folders not related to a Collection.
- Only allow editors and administrators to see those Folders.
- Filter files from orphan folders in search page for non editor or
  administrator users.

Other changes:

- Add `requests` requirement to connect with the SS and `vcrpy` to fake
  those requests in tests.
- Change Elasticsearch image in Docker environment.
- Ignore .tox folder in Docker Compose `watchmedo` command.
- Move search related tasks to search app.
- Remove some obvious comments and logs from tasks.

Related to #2 and #154.

jraddaoui commented:
Hi @cole, this is ready for review when you have some time.

First of all, sorry for mixing contents and a few back and forwards in this PR. I tried to detail it all in the description and on each commit but, if you think it's better to split this work in various PRs, I can try to clean it up a little.

A couple of important notes about this changes:

- Without entering in too much detail, two types of DIPs can be imported in the application. With this changes, the normal AM DIPs will work using the SS integration and the custom CCA DIPs will work using the Folder (DIP) creation form. Both DIPs should work on either workflow, but I want to check with CCA and the team before, to see how much we can normalize in the DIPs before making that work. That's why you'll see a few to-do comments in the tasks code.

- Initially, I decided to use `django_celery_results` to track task errors in the import process, which has turned out to be an inconvenient now that we're using task chains. I think you also had a conversation with Sevein about that not being a great idea, so we're not using them anymore. Nevertheless, I have not fully removed the `django_celery_results` requirement from the application as it involves migrations. I mentioned to CCA the unification of the core apps and the change in the project and migration names (from https://github.com/CCA-Public/dip-access-interface/pull/150#issuecomment-483364652), which may happen soon and it will ease the removal of `django_celery_results`.

Thanks!

jraddaoui commented:
A few more things @cole,

I have implement test for almost everything except the METS parsing process (`parsemets.py`), we have [an issue](https://github.com/CCA-Public/dip-access-interface/issues/73) to address that, as there are no tests for that entire module.

Also, this will require to add some documentation (for now to the read me file), but I think it's better to wait until the integration is "finished" for both DIP types.

cole commented on search/tasks.py:
```diff
@@ -0,0 +1,108 @@
+            "lang": "painless",
+            "params": {"collection": collection.get_es_data_for_files()},
+        }
+        files = DigitalFile.objects.filter(dip__collection__pk=pk).all()
```
I think you can use [.values_list('id', flat=True)](https://docs.djangoproject.com/en/2.1/ref/models/querysets/#values-list) here to save loading unnecessary fields.

cole commented on search/tasks.py:
```diff
@@ -0,0 +1,108 @@
+            "lang": "painless",
+            "params": data_params,
+        }
+        files = DigitalFile.objects.filter(dip__pk=pk).all()
```
values_list should also apply here.

cole commented on search/tasks.py:
```diff
@@ -0,0 +1,108 @@
+def delete_es_descendants(class_name, pk):
+    """Deletes the related documents in ES based on the ancestor id."""
+    if class_name not in ["Collection", "DIP"]:
+        raise Exception("Can not delete descendants of %s." % class_name)
```
```suggestion
        raise ValueError("Can not delete descendants of %s." % class_name)
```

jraddaoui commented on dips/api_views.py:
```diff
@@ -10,4 +15,49 @@ class DIPStoredWebhook(APIView):
+                dc=DublinCore.objects.create(identifier=dip_uuid),
+                import_status=DIP.IMPORT_PENDING,
+            )
+        else:
```
Oh, I didn't know about `defaults`, thanks!

jraddaoui commented on dips/models.py:
```diff
@@ -283,21 +270,17 @@ class DIP(AbstractEsModel):
-    # task is called from the `new_dip` view and to `SUCCESS` or `FAILURE` when
-    # the task ends from within the task's `after_return` method.
     import_status = models.CharField(max_length=7, null=True)
+    import_error_trace = models.TextField(blank=True, null=True)
```
Yes, these traces are shown to only some types of users in the Folder pages to provide some information about the error. But showing the error message instead of the entire trace should be enough.

jraddaoui commented on search/tasks.py:
```diff
@@ -0,0 +1,108 @@
+    default_retry_delay=30,
+    ignore_result=True,
+)
+def update_es_descendants(class_name, pk):
```
I had some conflicts about this too when I was doing it. Currently, the search tasks are called in the save and delete methods from the parent AbstractEsModel, which should make adding a new model with the same requirements easier. I should have moved the script declaration and other model/doc dependencies from those tasks to the actual model/doc to make it better, but I didn't expect that many differences and it ended all there ;(

About a single task to update each file (I'm not sure if that's what you're suggesting), I thought that, even in an async. context, it would be better to hit Elasticsearch as little as possible, using bulk requests.

jraddaoui commented on dips/views.py:
```diff
@@ -627,30 +624,93 @@ def delete_dip(request, pk):
+@login_required(login_url="/login/")
+def orphan_dips(request):
+    if not request.user.is_editor():
+        return redirect("home")
```
Actually, this is made in other views too, I'll create a new issue to properly use permissions, since we already have them defined.

jraddaoui replied:
https://github.com/CCA-Public/dip-access-interface/issues/155

jraddaoui commented:
Thanks a lot @cole!

I think I have addressed all except the permissions (for which I created #155) and the update descendants task (I added some comments above).
